crates/cxx-qt-lib/src/types/qvariant/mod.rs

```diff
@@ -0,0 +1,169 @@
+// SPDX-FileCopyrightText: 2021 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+// SPDX-FileContributor: Gerhard de Clercq <gerhard.declercq@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+use cxx::{type_id, ExternType};
+use std::mem::MaybeUninit;
+
+use crate::{
+    QColor, QDate, QDateTime, QPoint, QPointF, QRect, QRectF, QSize, QSizeF, QString, QTime, QUrl,
+};
+
+#[cxx::bridge]
+mod ffi {
+    unsafe extern "C++" {
+        include!("cxx-qt-lib/qvariant.h");
+        type QVariant = super::QVariant;
+
+        /// Convert this variant to type QMetaType::UnknownType and free up any resources used.
+        fn clear(self: &mut QVariant);
+        /// Returns true if this is a null variant, false otherwise.
+        #[rust_name = "is_null"]
+        fn isNull(self: &QVariant) -> bool;
+        /// Returns true if the storage type of this variant is not QMetaType::UnknownType; otherwise returns false.
+        #[rust_name = "is_valid"]
+        fn isValid(self: &QVariant) -> bool;
+    }
+
+    #[namespace = "rust::cxxqtlib1"]
+    unsafe extern "C++" {
+        include!("cxx-qt-lib/common.h");
+
+        #[doc(hidden)]
+        #[rust_name = "qvariant_drop"]
+        fn drop(variant: &mut QVariant);
+        #[doc(hidden)]
+        #[rust_name = "qvariant_default"]
+        fn construct() -> QVariant;
+        #[doc(hidden)]
+        #[rust_name = "qvariant_clone"]
+        fn construct(variant: &QVariant) -> QVariant;
+    }
+}
+
+/// The QVariant class acts like a union for the most common Qt data types.
+#[repr(C)]
+pub struct QVariant {
+    /// The layout has changed between Qt 5 and Qt 6
+    ///
+    /// Qt5 QString has one member, which contains three uints (but they are optimised to a size of 8) and a union
+    /// Qt6 QString has one member, which contains three pointers and a union (pointer largest)
+    #[cfg(qt_version_major = "5")]
+    _space: MaybeUninit<[usize; 2]>,
+    #[cfg(qt_version_major = "6")]
+    _space: MaybeUninit<[usize; 4]>,
+}
+
+impl Clone for QVariant {
+    /// Constructs a copy of the variant, p, passed as the argument to this constructor.
+    fn clone(&self) -> Self {
+        ffi::qvariant_clone(self)
+    }
+}
+
+impl Default for QVariant {
+    /// Constructs an invalid variant.
+    fn default() -> Self {
+        ffi::qvariant_default()
+    }
+}
+
+impl Drop for QVariant {
+    /// Destroys the QVariant and the contained object.
+    fn drop(&mut self) {
+        ffi::qvariant_drop(self)
+    }
+}
+
+impl<T> From<&T> for QVariant
+where
+    T: QVariantValue,
+{
+    /// Constructs a QVariant from a value of T
+    fn from(value: &T) -> Self {
+        T::construct(value)
+    }
+}
+
+// Note we can't use impl Into or TryInto for QVariant here as it conflicts
+//
+// note: conflicting implementation in crate `core`:
+// - impl<T, U> TryInto<U> for T
+//   where U: TryFrom<T>;
```

Comment:
So the fun is that you can't do `impl<T> From<&QVariant> for T where T: QVariantValue` as you get `type parameter T must be covered by another type when it appears before the first local type (`qvariant::QVariant`)`. So then before i'm manually implemented `Into` to just get `.into()` rather than from+into. But you can't do that as it collides.

It seems that the use of generics/traits confuses things :-/ As ideally i wanted to be able to go `variant.into()` or `variant.try_into()`, to do the conversion from a `QVariant` to a `T`. So then the closest we can get is having methods that do that for us on the `QVariant` (the `try_value` and `value`) which seem to work well.

Reply:
So i was going to leave the comment there to remind that it isn't possible to do this, even then it would be if it wasn't `T: QVariantValue`